M2/Macaulay2/packages/Msolve.m2

```diff
@@ -0,0 +1,527 @@
+    inStr:=l1|newline|l2|newline|Igens;
+    mIn<<inStr<<close;
+    mOut:=temporaryFileName()|".ms";
+    << "msolve input file is called: " << mIn << endl;
+    << "msolve output file is called: " << mOut << endl;
+    callStr:=msolveEXEwithOpts(opt)|" -g 2 -f "|mIn|" -o "|mOut;
+    run(callStr);
+    <<"done msolve"<<endl;
+    msolGB:=readMsolveOutputFile(R, mOut);
+    return gens forceGB msolGB;
+    );
+msolveLeadMonomials=method(TypicalValue=>Matrix, Options=>{"level of verbosity"=>0, "number of threads"=>maxAllowableThreads});
+msolveLeadMonomials(Ideal):=opt->(I)->(
+    if not inputOkay(I) then return 0;
+    mIn:=temporaryFileName()|".ms";
+    R:=ring I;
+    kk:=coefficientRing(R);
+    gR:=toString(gens R);
+    l1:=substring(1,#gR-2,gR);
+    l2:=char R;
+    gI:=toString flatten entries gens I;
+    if (isField(kk) and (char(kk)==0)) then gI=replace("[)(]","",gI);
+    Igens:=replace(" ",""|newline,substring(1,#gI-2,gI));
+    inStr:=l1|newline|l2|newline|Igens;
+    mIn<<inStr<<close;
+    mOut:=temporaryFileName()|".ms";
+    callStr:=msolveEXEwithOpts(opt)|" -g 1 -f "|mIn|" -o "|mOut;
+    run(callStr);
+    msolGB:=readMsolveOutputFile(R, mOut);
+    return gens forceGB msolGB;
+    );
+msolveEliminate=method(Options=>{"level of verbosity"=>0, "number of threads"=>maxAllowableThreads});
+msolveEliminate(Ideal,RingElement):=Ideal =>opt-> (I,elimvar)->(
+    return msolveEliminate(I,{elimvar});
+    );
+msolveEliminate(RingElement,Ideal):=Ideal => opt->(elimvar,I) ->
+    msolveEliminate(I,{elimvar})
+msolveEliminate(List,Ideal):=Ideal => opt->(elimvars,I) -> 
+    msolveEliminate(I,elimvars)
+msolveEliminate(Ideal,List):=Ideal => opt->(J,elimvars)->(
+    if not inputOkay(J) then return 0;
+    mIn:=temporaryFileName()|".ms";
+    S:=ring J;
+    kk:=coefficientRing(S);
+    keepvars:=select(gens S, s->not member(s,elimvars)); 
+    elimNum:=length(elimvars);
+    --R:=kk(monoid [join(elimvars,keepVars),MonomialOrder=>{numgens(S):1}]);
+    R:=kk(monoid [join(elimvars,keepvars),MonomialOrder=>{#elimvars, #keepvars}]);
+    elimR:=kk(monoid [keepvars]);
+    I:=sub(J,R);
+    gR:=toString(gens R);
+    l1:=substring(1,#gR-2,gR);  -- removing parentheses
+    l2:=char R;
+    gI:=toString flatten entries gens I;
+    if (isField(kk) and (char(kk)==0)) then gI=replace("[)(]","",gI);
+    Igens:=replace(" ",""|newline,substring(1,#gI-2,gI));
+    inStr:=l1|newline|l2|newline|Igens;
+    mIn<<inStr<<close;
+    mOut:=temporaryFileName()|".ms";
+    << "mOut = " << mOut << endl;
+    callStr:=msolveEXEwithOpts(opt)|" -e "|toString(elimNum)|" -g 2 -f "|mIn|" -o "|mOut;
+    run(callStr);
+    if char R === 0 then 
+      ideal readMsolveOutputFile(elimR, mOut)
+    else
+      ideal readMsolveOutputFile(R, mOut)
+    )
+
+msolveSaturate=method(TypicalValue=>Matrix, Options=>{"level of verbosity"=>0, "number of threads"=>maxAllowableThreads});
+msolveSaturate(Ideal,RingElement):=opt->(I,f)->(
+    if not inputOkay(I) then return 0;
```

Comment:
I would want this to be turned into a hook/strategy for `saturate` the most, and if it threw an error instead it would be very easy to do that:
```m2
addHook((saturate, Ideal, RingElement), msolveSaturate, Strategy => MSolve)
```
That's it!